core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java

```diff
@@ -297,6 +297,84 @@ public void testGreaterThanOrEqual()
         testUnwrap("bigint", "a >= DOUBLE '-18446744073709551616'", "NOT (a IS NULL) OR NULL");
     }
 
+    @Test
+    public void testBetween()
+    {
+        // representable
+        testUnwrap("smallint", "a BETWEEN DOUBLE '1' AND DOUBLE '2'", "a BETWEEN SMALLINT '1' AND SMALLINT '2'");
+        testUnwrap("bigint", "a BETWEEN DOUBLE '1' AND DOUBLE '2'", "a BETWEEN BIGINT '1' AND BIGINT '2'");
+        testUnwrap("decimal(7, 2)", "a BETWEEN DOUBLE '1.23' AND DOUBLE '4.56'", "a BETWEEN CAST(DECIMAL '1.23' AS decimal(7, 2)) AND CAST(DECIMAL '4.56' AS decimal(7, 2))");
+        // cast down is possible
+        testUnwrap("decimal(7, 2)", "CAST(a AS DECIMAL(12,2)) BETWEEN CAST(DECIMAL '111.00' AS decimal(12,2)) AND CAST(DECIMAL '222.0' AS decimal(12,2))", "a BETWEEN CAST(DECIMAL '111.00' AS decimal(7, 2)) AND CAST(DECIMAL '222.00' AS decimal(7, 2))");
+
+        // non-representable, min cast round down, max cast round down
+        testUnwrap("bigint", "a BETWEEN DOUBLE '1.1' AND DOUBLE '2.2'", "a BETWEEN BIGINT '2' AND BIGINT '2'");
+        // non-representable, min cast round down, max cast round up
+        testUnwrap("bigint", "a BETWEEN DOUBLE '1.1' AND DOUBLE '1.9'", "a > BIGINT '1' AND a < BIGINT '2'");
+        // non-representable, min cast round down, max cast no rounding
+        testUnwrap("bigint", "a BETWEEN DOUBLE '1.1' AND DOUBLE '2'", "a BETWEEN BIGINT '2' AND BIGINT '2'");
+        // non-representable, min cast round up, max cast round down
+        testUnwrap("bigint", "a BETWEEN DOUBLE '1.9' AND DOUBLE '2.2'", "a BETWEEN BIGINT '2' AND BIGINT '2'");
```

Comment:
This `BETWEEN` predicate can be turned into an equality.
I didn't do it yet because the PR is currently just a draft.